plugin/kubernetes/external.go

```diff
@@ -96,27 +142,54 @@ func (k *Kubernetes) External(state request.Request) ([]msg.Service, int) {
 }
 
 // ExternalAddress returns the external service address(es) for the CoreDNS service.
-func (k *Kubernetes) ExternalAddress(state request.Request) []dns.RR {
+func (k *Kubernetes) ExternalAddress(state request.Request, headless bool) []dns.RR {
 	// If CoreDNS is running inside the Kubernetes cluster: k.nsAddrs() will return the external IPs of the services
 	// targeting the CoreDNS Pod.
 	// If CoreDNS is running outside of the Kubernetes cluster: k.nsAddrs() will return the first non-loopback IP
 	// address seen on the local system it is running on. This could be the wrong answer if coredns is using the *bind*
 	// plugin to bind to a different IP address.
-	return k.nsAddrs(true, state.Zone)
+	return k.nsAddrs(true, headless,  state.Zone)
 }
 
-// ExternalServices returns all services with external IPs
-func (k *Kubernetes) ExternalServices(zone string) (services []msg.Service) {
+// ExternalServices returns all services with external IPs and if enabled headless services
+func (k *Kubernetes) ExternalServices(zone string, headless bool) (services []msg.Service) {
 	zonePath := msg.Path(zone, coredns)
 	for _, svc := range k.APIConn.ServiceList() {
-		for _, ip := range svc.ExternalIPs {
-			for _, p := range svc.Ports {
-				s := msg.Service{Host: ip, Port: int(p.Port), TTL: k.ttl}
-				s.Key = strings.Join([]string{zonePath, svc.Namespace, svc.Name}, "/")
-				services = append(services, s)
-				s.Key = strings.Join(append([]string{zonePath, svc.Namespace, svc.Name}, strings.ToLower("_"+string(p.Protocol)), strings.ToLower("_"+string(p.Name))), "/")
-				s.TargetStrip = 2
-				services = append(services, s)
+		// Endpoint query or headless service
+		if headless && svc.Headless() {
+			
+			idx := object.ServiceKey(svc.Name, svc.Namespace)
+		    endpointsList :=  k.APIConn.EpIndex(idx)
+	
+			for _, ep := range endpointsList {
+				if ep.Name != svc.Name || ep.Namespace != svc.Namespace {
+					continue
+				}
+
+				for _, eps := range ep.Subsets {
+					
+					for _, addr := range eps.Addresses {
+
+						for _, p := range eps.Ports {
+							s := msg.Service{Host: addr.IP, Port: int(p.Port), TTL: k.ttl}
+							baseSvc := strings.Join([]string{zonePath, svc.Namespace, svc.Name}, "/")
+							s.Key = strings.Join([]string{baseSvc, endpointHostname(addr, k.endpointNameMode)}, "/")
+							services = append(services, s)
+						}
+					}
+				}
+			}
+			continue
```

Comment:
Also need to append the headless service record (without endpoint label) to `services`.

Reply:
I think that it's added in previous section now. :slightly_smiling_face: